modules/substrate/src/lib.rs

```diff
@@ -105,53 +105,18 @@ decl_storage! {
 		// Grandpa doesn't require there to always be a pending change. In fact, most of the time
 		// there will be no pending change available.
 		NextScheduledChange: map hasher(identity) BridgedBlockHash<T> => Option<ScheduledChange<BridgedBlockNumber<T>>>;
+		/// Whether or not the bridge has been initialized.
+		///
+		/// This is important to know to ensure that we don't try and initialize the bridge twice
+		/// and create an inconsistent genesis state.
+		IsInitialized: bool;
```

Comment:
I would rather remove all the is-already-initialized checks at all. Root is here, whether we want this, or not. Instead of calling `initialize()`, root may just update storage values directly, without calling the function. Why would we need to build gate in the middle of the unfenced field? I agree that in "normal programming", we should protect against double-initialization. But here it only takes additional resources - no matter if we read `IsInitialized` or ``CurrentAuthoritySet.

I could even imagine this function evolved to the `reinitialize` that may be called if we have missed too many headers and just want to restart the sync.

Reply:
> I'm assuming that you're going to reply that this feels hacky.

You know me too well 😄 

> Since the uninitialized code is hacky already, cause the invariants are broken (for instance ImportedHash does not contain BestFinalized

I'm not entirely sure what you mean here

> Next I'd probably pass since there is no good solution here, for safety it would be good to do initialization check for each dispatchable (and maybe even a runtime call), but it might not be worth the storage read costs.

Let me propose a different solution here. Since we're going to be introducing halting to the pallet, all the dispatchables are going to be checking if the pallet is halted before proceeding. The initial state of the pallet needs to be set as a configuration parameter (either via `GenesisConfig` or `initialize()`). We can enforce that if the pallet is not initialized then it must be halted. Once we initialize the pallet then we can lift the halted state (or keep it halted if that's what's desired).

This would have to be done in the halting PR. What do you think?

Reply:
I've followed up on this in  #485.